File: crates/app/ui/src/lib/sidebar.ts
// The Sidebar tab's pre-load default, the process-threshold option sets, and the usage floor each
// threshold implies. Each threshold is a closed enum (mirrors core); the panel maps a variant to
// its display label, and a process row shows its CPU/memory read-out only once usage reaches the
// mapped floor — both mappings live here, in one place.

import type { Option } from "@/lib/appearance";
import type { ProcessCpuThreshold, ProcessMemThreshold, Sidebar } from "@/domain";

// Mirrors core::Sidebar::default(): filter on, empty sections shown, usage always shown, footer on.
// The facade's stored value supersedes this on load.
export const DEFAULT_SIDEBAR: Sidebar = {
  show_filter_input: true,
  hide_empty_sections: false,
  process_cpu_threshold: "always",
  process_mem_threshold: "always",
  show_settings_footer: true,
};

export const PROCESS_CPU_OPTIONS: Option<ProcessCpuThreshold>[] = [
  { value: "always", label: "Always" },
  { value: "pct10", label: "10%" },
  { value: "pct30", label: "30%" },
  { value: "pct60", label: "60%" },
  { value: "pct90", label: "90%" },
  { value: "never", label: "Never" },
];

export const PROCESS_MEM_OPTIONS: Option<ProcessMemThreshold>[] = [
  { value: "always", label: "Always" },
  { value: "mb100", label: "100 MB" },
  { value: "mb500", label: "500 MB" },
  { value: "gb1", label: "1 GB" },
  { value: "gb2", label: "2 GB" },
  { value: "never", label: "Never" },
];

const MIB = 1024 * 1024;
const GIB = 1024 * MIB;

// The CPU-percent a process must reach for its row to show the CPU read-out. `always` shows at any
// usage (floor 0); `never` hides it (floor Infinity).
export const PROCESS_CPU_FLOOR: Record<ProcessCpuThreshold, number> = {
  always: 0,
  pct10: 10,
  pct30: 30,
  pct60: 60,
  pct90: 90,
  never: Infinity,
};

// The resident bytes a process must reach for its row to show the memory read-out.
export const PROCESS_MEM_FLOOR: Record<ProcessMemThreshold, number> = {
  always: 0,
  mb100: 100 * MIB,
  mb500: 500 * MIB,
  gb1: GIB,
  gb2: 2 * GIB,
  never: Infinity,
};

File: crates/core/src/settings/sidebar.rs
//! Sidebar settings (global Sidebar tab): what the process-tree sidebar shows — the filter input,
//! empty-section hiding, the per-process CPU/memory usage thresholds, and the settings footer.
//!
//! Each usage threshold is a closed enum (the discrete options the demo offers), so a process row
//! shows its CPU/memory read-out only once usage reaches the chosen level — never a bare percentage
//! or byte count. The frontend maps each variant to its display label and comparison value.

use serde::{Deserialize, Serialize};

/// When a process row shows its CPU-usage read-out.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProcessCpuThreshold {
    #[default]
    Always,
    Pct10,
    Pct30,
    Pct60,
    Pct90,
    Never,
}

/// When a process row shows its memory-usage read-out.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProcessMemThreshold {
    #[default]
    Always,
    Mb100,
    Mb500,
    Gb1,
    Gb2,
    Never,
}

/// The Sidebar tab document. Every field carries a serde default so an older record still reads.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default)]
pub struct Sidebar {
    /// Show the filter input at the top of the sidebar.
    pub show_filter_input: bool,
    /// Hide sections with no processes (e.g. Agents, Terminals).
    pub hide_empty_sections: bool,
    /// When a process row shows its CPU usage.
    pub process_cpu_threshold: ProcessCpuThreshold,
    /// When a process row shows its memory usage.
    pub process_mem_threshold: ProcessMemThreshold,
    /// Show the Settings button at the bottom of the sidebar (still reachable via palette and hotkey).
    pub show_settings_footer: bool,
}

impl Default for Sidebar {
    fn default() -> Self {
        Self {
            show_filter_input: true,
            hide_empty_sections: false,
            process_cpu_threshold: ProcessCpuThreshold::default(),
            process_mem_threshold: ProcessMemThreshold::default(),
            show_settings_footer: true,
        }
    }
}
